classes/04class/iac/README.md

```diff
@@ -0,0 +1,304 @@
+
+On the variable declaration, you can specify a default value, so if the value is not included during the code execution, it will use the default one.
+
+## Outputs
+Outputs are like return values from a terraform code. It can be useful, so you receive an attribute of a resource being created. It can also be used by another piece of code outside of your own.
+
+```terraform
+output "instance_ip_addr" {
+  value = aws_instance.server.private_ip
+}
+```
+The piece of code above would print, after executing the Terraform code, the private IP address of the instance created.
+
+Each kind of resource has specific attributes that can be output, and you can check those attributes on the provider resource page.
+
+## File Organization
+Just like any other programming language, Terraform has it's best practices to organize the code you create.
+
+Everything (resources, outputs, providers, state, etc) could be written in a single file with a `.tf` extension, however, to make things easier, each kind of code block should be created on a specific file.
+- **_provider.tf**
+  - All providers are declared on this file
+- **_backend.tf**
+  - The backend to be used is declared on this file
+- **variables.tf**
+  - The variables expected to be received during the code execution are declared on this file.
+- **_outputs.tf**
+  - All outputs from your code are declared on this file
+- **resource_related.tf**
+  - For the resources, you'll normally create a single file for each `family` of resources. Normally you'll have an `iam.tf`, `vpc.tf`, `s3.tf`, `rds.tf`, etc. In this example, all resources related to IAM, would go into the `iam.tf` file.
+- **anything.tfvars**
+  - In this file you would have the variables to be used during the code execution. If you have multiple environments, you can have your `dev.tfvars` and `prod.tfvars` files containing specific values for each environment and point to each file during execution based on the environment you want to apply your code.
+
+## Running some code
+Now that we have some understanding about terraform and its power let's give it a go and run some code.
+
+First of all, make sure you have your AWS credentials configured on your machine. You can use the guide from [here](https://link-to-the-cli-config)
+
+All the code we'll use is on the [artifacts folder](artifacts/terraform) of this repository, so just get into that directory. That way terraform will consider it as the working directory.
+
+```bash
+# Considering that you are on the folder of this file
+$ cd artifacts/terraform
+```
+Let's initialize terraform
+```bash
+$ terraform init
+```
+Now that terraform is initialized, we can run the next command, which would determine the resources that need to be created/updated/removed.
+```bash
+$ terraform plan -var-file=./class.tfvars
```

Comment:
do you think we should use manual tfvars files? Wouldn't be easier at this stage if we just stick to normal tf files? Might be nice to still have something about tfvars and how can  you run the same code with multiple var files but as an extra documentation, not for the class

Reply:
I don't see a lot of difference between using a external var with tfvars or just a local var in the code.
The sense is the same, but the latter is much more useful, so tried to explore more that option instead of the local vars.

Reply:
just one more complexity added for something that is already quite hard to understand

Reply:
how harder is tfvars instead of locals?
Or you reckon we should avoid variables at all?

Reply:
reached an agreement to keep the variables the way it is.